odoo/sql_db.py

```diff
@@ -303,6 +303,12 @@ def process(type):
 
     @check
     def close(self):
+        # Call rollback hooks if some exists.
+        # This list will be empty if the commit method has been called
+        # before closing te cursor
+        for func in self._pop_event_handlers()['rollback']:
+            func()
```

Comment:
I suggest to move that code at the very end of method `_close()`, i.e., *after* the call `self._cnx.rollback()`.  This is more in line with the expected behavior (after rollback), and guarantees that the hooks are processed even when `_close()` is called by the garbage collector...

Reply:
ok done